feat: merge account updates details

Mirko-von-Leipzig opened the pull request:
This PR adds `AccountUpdateDetails::merge` which combines two update details. This is required by the node to [support merging multiple transactions](https://github.com/0xPolygonMiden/miden-node/pull/407#discussion_r1679166952) on the same account.

~I implemented the merging using hashmaps to ensure only the final update is applied. This required `Hash` on several structs, and unfortunately `Felt` does not implement `Hash` itself so this required manually implementing it in some cases.~ Use `BTreeMap` instead which doesn't require `Hash` and is `nostd` compliant.

The structs themselves actually look like they should contain hashmaps instead of vectors as this would be more type safe imo. However, since they aren't this makes me question my approach. I'll ask more specific questions inline so that the conversation is easier to track.

~I also have some `todo!()` which I don't know how to implement.~

As a side-note we could also use something like `Slot` type, similar to `NonZeroU8` but which disallows the "illegal" slots.

Mirko-von-Leipzig commented on objects/src/accounts/delta/storage.rs:
```diff
@@ -40,6 +42,32 @@ impl AccountStorageDelta {
+        Self {
+            cleared_items,
+            updated_items,
+            updated_maps: todo!(),
```
I don't quite understand maps and how they're different from normal updates. afaiu they're a way of getting more than 254 slots? However I don't know how that is implemented.

In the validate fn there is also a comment about ensuring that every map update should have a corresponding item update, however I don't see any such check.

bobbinth replied:
@Dominik1999 can provide additional info on how storage works.

Dominik1999 replied:
So, `StorageMaps` are sparse Merkle trees of depth 64 with `Word`s as key-value pairs. Users can store a lot of data using `StorageMaps`. 

The root of a map is stored in a `StorageItem`. So for every change in the map, the corresponding item also changes because the root changes. 

Does that help?

Mirko-von-Leipzig replied:
Thanks. The current implementation is 
```rust
pub struct AccountStorageDelta {
    pub cleared_items: Vec<u8>,
    pub updated_items: Vec<(u8, Word)>,
    pub updated_maps: Vec<(u8, StorageMapDelta)>,
}
```
I assume the `u8` in `updated_maps` represents the `StorageItem` slot then? If so, then if I want to merge two delta's then I need to be ensure that a normal `updated_item` does not clash with an `updated_map` item?

bobbinth replied:
Yes, if a delta is valid, there should be no overlap between updated/cleared simple slots and updated maps.

This does bring up one point though: this function assumes that both deltas are valid. We should probably mention this is in the doc comment to make the assumption explicit.

We could also go anything beyond that - i.e., validate the incoming deltas and either panic or return an error if either of them is not valid. But I'm not sure that's needed.

Dominik1999 replied:
There should be a check for the `StorageLayout` that defines which slots are maps and which items (= `Value`) 

For example [here](https://github.com/0xPolygonMiden/miden-base/blob/13724e96ee8a423a192737bb6a7962cfcea8b792/objects/src/accounts/storage/mod.rs#L311C12-L335) we assert only a item of type `Value` can be set using `set_item`. A map cannot be set here.

Mirko-von-Leipzig commented on objects/src/accounts/delta/vault.rs:
```diff
@@ -37,6 +39,32 @@ impl AccountVaultDelta {
+        //
+        // Ordering matters. Since later items will overwrite earlier ones we need to begin
+        // with self, then other.
+        let assets = self
+            .added_assets
+            .into_iter()
+            .map(|asset| (asset, true))
+            .chain(self.removed_assets.into_iter().map(|asset| (asset, true)))
+            .chain(other.added_assets.into_iter().map(|asset| (asset, false)))
+            .chain(other.removed_assets.into_iter().map(|asset| (asset, false)))
+            .collect::<HashMap<_, _>>();
```
I'm wondering why it isn't something like

```rust
pub struct AccountVaultDelta(HashMap<Asset, AssetDelta>);

enum AssetDelta {
  Added,
  Removed,
}
```

bobbinth replied:
I don't think there is a strong reason - probably was done like this to more closely follow the serialization format. Could you create an issue to consider refactoring this?

Mirko-von-Leipzig replied:
https://github.com/0xPolygonMiden/miden-base/issues/798

Mirko-von-Leipzig commented on objects/src/accounts/delta/storage.rs:
```diff
@@ -40,6 +42,32 @@ impl AccountStorageDelta {
     }
 
+    /// Merges another delta into this one, overwriting any existing values.
+    pub fn merge(self, other: Self) -> Self {
```
Similar question regarding using `HashMap<slot, value|removed|map>` as an implementation.

Mirko-von-Leipzig replied:
https://github.com/0xPolygonMiden/miden-base/issues/798

bobbinth commented on objects/src/accounts/delta/mod.rs:
```diff
@@ -58,6 +58,15 @@ impl AccountDelta {
+    /// Merge another [AccountDelta] into this one.
+    pub fn merge(self, other: Self) -> Result<Self, AccountDeltaError> {
+        // Incoming nonce takes precedence.
+        let nonce = other.nonce.or(self.nonce);
```
I would add a check here that incoming nonce is greater than current nonce.

Mirko-von-Leipzig replied:
I've added the check. I'd like to add some derives to `Felt` - `PartialOrd`, `Ord` so that these comparisons are less clunky.

Mirko-von-Leipzig commented on objects/src/accounts/account_id.rs:
```diff
@@ -104,6 +104,12 @@ pub enum AccountStorageType {
 #[cfg_attr(feature = "serde", serde(transparent))]
 pub struct AccountId(Felt);
 
+impl std::hash::Hash for AccountId {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        self.0.inner().hash(state)
+    }
+}
```
Ah perfect; I always overlook `BTreeMap` for some reason.

Mirko-von-Leipzig commented on objects/src/assets/nonfungible.rs:
```diff
@@ -26,6 +26,28 @@ const FAUCET_ID_POS: usize = 1;
 
+impl PartialOrd for NonFungibleAsset {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        /// This wrapper allows us to use iterators builtin partial ord implementation.
+        /// This is much safer than attempting to do this correctly.
+        #[derive(PartialEq, Eq)]
+        struct Helper<'a>(&'a Felt);
+        impl PartialOrd for Helper<'_> {
+            fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+                self.0.inner().partial_cmp(&other.0.inner())
+            }
+        }
+
+        self.0.iter().map(Helper).partial_cmp(other.0.iter().map(Helper))
```
We should derive `PartialOrd` and `Ord` for `Felt` if that's possible.

bobbinth replied:
I've been trying to avoid doing that because technically field elements do not inherently have an order. Their integer representation does have order - so, if we need to compare field elements, we usually first convert them to integers (i.e., via `felt.as_int()` method).

bobbinth replied:
For this specific case, I think converting `Word` into a `Digest` (which should be a zero-cost conversion) and then comparing digests (which implement `PartialOrd`, I believe), could be a simpler solution.

Mirko-von-Leipzig replied:
Could you link me `Digest`? I'm struggling to find one that isn't the trait. I think technically its also not zero-cast because we're working with a reference here so it would copy/clone it?

bobbinth replied:
The struct is actually called `RpoDigest` and is defined in `miden-crypto` but in this crate we import it under `Digest` alias [here](https://github.com/0xPolygonMiden/miden-base/blob/main/objects/src/lib.rs#L32).

bobbinth commented on objects/src/accounts/delta/storage.rs:
```diff
@@ -40,6 +40,32 @@ impl AccountStorageDelta {
         }
     }
 
+    /// Merges another delta into this one, overwriting any existing values.
+    pub fn merge(self, other: Self) -> Self {
```
I would add some tests to make sure this works correctly.

bobbinth commented on objects/src/accounts/delta/vault.rs:
```diff
@@ -37,6 +37,28 @@ impl AccountVaultDelta {
 
+    /// Merges another delta into this one, overwriting any existing values.
+    pub fn merge(self, other: Self) -> Self {
+        // Collec the assets into a hashmap where true indicates added, and false removed.
+        // This lets us simplify the addition/removal dance we would otherwise have to go through.
+        //
+        // Ordering matters. Since later items will overwrite earlier ones we need to begin
+        // with self, then other.
+        let assets = self
+            .added_assets
+            .into_iter()
+            .map(|asset| (asset, true))
+            .chain(self.removed_assets.into_iter().map(|asset| (asset, true)))
+            .chain(other.added_assets.into_iter().map(|asset| (asset, false)))
+            .chain(other.removed_assets.into_iter().map(|asset| (asset, false)))
+            .collect::<BTreeMap<_, _>>();
```
We should update the comment to mention `BTreeMap` instead of hashmap, but also - I'm not sure this logic is correct. Specifically, I'm not sure it will merge fungible assets correctly.

For example, if `self` has added asset `(account_id1, 100)` and `other` has added asset `(account_id1, 200)`, as far as I can tell, these would not be merged into a single asset `(account_id1, 300)`.

I think we probably should process fungible and non-fungible assets separately here. For non-fungible assets, something like the above logic should work. For fungible assets, we probably need something like `BTreeMap<AccountId, amount>` where amount could be either positive or negative (i.e., we could use `i64`).

bobbinth commented on objects/src/accounts/delta/vault.rs:
```diff
@@ -37,6 +37,28 @@ impl AccountVaultDelta {
         }
     }
 
+    /// Merges another delta into this one, overwriting any existing values.
+    pub fn merge(self, other: Self) -> Self {
```
I would add some tests for this function.

bobbinth commented on objects/src/assets/fungible.rs:
```diff
@@ -12,7 +12,7 @@ use super::{
 /// A fungible asset consists of a faucet ID of the faucet which issued the asset as well as the
 /// asset amount. Asset amount is guaranteed to be 2^63 - 1 or smaller.
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
 pub struct FungibleAsset {
```
Given one of the above comments, I'm not sure this change would be necessary.

bobbinth reviewed (commented):
Thank you! Looks good. I left some comments inline. Also, I think we need to finish implementing merging of `StorageMapDelta` structs.

Mirko-von-Leipzig commented on objects/src/accounts/delta/mod.rs:
```diff
@@ -104,6 +120,28 @@ impl AccountUpdateDetails {
+            (AccountUpdateDetails::Delta(initial), AccountUpdateDetails::Delta(new_delta)) => {
+                AccountUpdateDetails::Delta(initial.merge(new_delta)?)
+            },
+            _ => todo!("Illegal combination error"),
```
Will do; just waiting on all the merge parts to complete so I know what errors I'm dealing with

bobbinth commented on objects/src/accounts/delta/vault.rs:
```diff
@@ -37,6 +41,99 @@ impl AccountVaultDelta {
         }
     }
 
+    /// Merges another delta into this one, overwriting any existing values.
+    ///
+    /// Inputs and the result are validated as part of the merge.
+    pub fn merge(self, other: Self) -> Result<Self, AccountDeltaError> {
```
Looking at how painful this is, I very much agree that we should refactor the delta to look something like this:
```rust
pub struct AccountVaultDelta {
    fungible: BTreeMap<AccountId, i64>,
    non_fungible: BTreeMap<NonFungibleAsset, bool>,
}
```
But since this is already working, we can probably leave the refactoring to #798.

One potential caveat that the ordering of assets may become important in the future - but we can cross that bridge when we get there.

bobbinth commented on objects/src/accounts/delta/vault.rs:
```diff
@@ -37,6 +41,99 @@ impl AccountVaultDelta {
+    ///
+    /// Inputs and the result are validated as part of the merge.
+    pub fn merge(self, other: Self) -> Result<Self, AccountDeltaError> {
+        self.validate()?;
+        other.validate()?;
```
Is there a reason to validate inputs/output when merging `AccountValutDelta` but not for other deltas (e.g., `AccountStorageDelta`)?

Mirko-von-Leipzig replied:
I think we may as well. I originally thought we could just leave it to the caller; but the more I implement the less certain I am of correctness - so adding additional checks is probably good.

I think a good goal when this is refactored is to enforce safe-by-construction alongside it.

bobbinth reviewed (commented):
Not a full review - but I did leave a few more comments/questions inline.

Mirko-von-Leipzig commented:
I believe I've addressed all review comments. I've added tests - though note that I used `rstest` to reduce boiler-plate. In general I find this to be a useful dependency but I can understand if you prefer I write the cases out instead.

bobbinth commented on objects/src/errors.rs:
```diff
@@ -63,6 +66,8 @@ pub enum AccountDeltaError {
     TooManyRemovedAssets { actual: usize, max: usize },
     TooManyUpdatedStorageItems { actual: usize, max: usize },
     DuplicateStorageMapLeaf { key: RpoDigest },
+    AmountTooBig(u64),
```
nit: I would probably make the naming a bit more explicit - e.g., `AssetAmountTooBig`.

bobbinth commented on objects/src/assets/fungible.rs:
```diff
@@ -12,7 +12,7 @@ use super::{
 /// A fungible asset consists of a faucet ID of the faucet which issued the asset as well as the
 /// asset amount. Asset amount is guaranteed to be 2^63 - 1 or smaller.
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
```
I think this change should probably be reverted?

bobbinth reviewed (approved):
Looks good! Thank you! I left just a couple of minor comments inline. Once these are addressed, we can merge.
